daphne/src/roles/helper.rs

```diff
@@ -216,42 +217,104 @@ pub trait DapHelper<S>: DapAggregator<S> {
+
+            for t in agg_job_resp.transitions {
+                for (id, var) in &mut output_transitions {
+                    if *id == t.report_id {
+                        *var = Some(t.var);
+                        break;
+                    }
                 }
-                retry_count -= 1;
-                let (agg_share_span, agg_job_resp) = task_config.vdaf.handle_agg_job_cont_req(
-                    task_id,
-                    task_config,
-                    &state,
-                    |id| replayed_reports.contains(id),
-                    &agg_job_id,
-                    &agg_job_cont_req,
-                    &metrics,
-                )?;
-
-                let out_shares_count = agg_share_span.report_count().try_into().unwrap();
-                let replayed = self
-                    .try_put_agg_share_span(task_id, task_config, agg_share_span)
-                    .await?;
-
-                if let Some(replayed) = replayed {
-                    replayed_reports.extend(replayed);
-                    metrics.agg_job_cont_restarted_inc();
-                } else {
-                    break (out_shares_count, agg_job_resp);
+            }
+
+            let inc_restart_metric = Once::new();
+            for (_, result) in put_shares_result {
+                match result {
+                    // we got 0 replays
+                    (Ok(replays), reports) if replays.is_empty() => {
+                        // remove successfull reports in this batch as they are finished
+                        input_transitions
+                            .retain(|t| reports.iter().all(|(id, _)| *id != t.report_id));
+                        // set those reports as finished in the output_transitions
+                        output_transitions
+                            .iter_mut()
+                            .filter(|(id, _)| reports.iter().any(|id_time| id_time.0 == *id))
+                            .for_each(|(_, var)| *var = Some(TransitionVar::Finished));
+                    }
+                    // this bucket had replays
+                    (Ok(replays), _) => {
+                        // remove the replays from the input transitions
+                        input_transitions.retain(|t| !replays.contains(&t.report_id));
+                        // mark the replayed reports as replays in the output_transitions
+                        output_transitions
+                            .iter_mut()
+                            .filter(|(id, _)| replays.contains(id))
+                            .for_each(|(_, var)| {
+                                *var = Some(TransitionVar::Failed(
+                                    crate::messages::TransitionFailure::ReportReplayed,
+                                ))
+                            });
+                        inc_restart_metric.call_once(|| metrics.agg_job_cont_restarted_inc());
+                    }
+                    (Err(e), _) => return Err(e.into()),
```

Comment:
Worth commenting that failure here might result in an a botched aggregate result (Leader and Helper may have inconsistent aggregate shares).